[DDP] Log if graph is static at end of training

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #61871

When set_static_graph=False, the only type of dynamism we really
support in DDP is dynamic set of unused parameters which must be explicitly
enabled with find_unused_parameters=True. Although, some workflows have static
set of unused parameters, would be good to detect and add this to logging to
identify workflows that are candidates for static graph optimization.

Differential Revision: [D29773962](https://our.internmc.facebook.com/intern/diff/D29773962/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D29773962/)!

rohan-varma committed:
[DDP] Log if graph is static at end of training

When set_static_graph=False, the only type of dynamism we really
support in DDP is dynamic set of unused parameters which must be explicitly
enabled with find_unused_parameters=True. Although, some workflows have static
set of unused parameters, would be good to detect and add this to logging to
identify workflows that are candidates for static graph optimization.

Differential Revision: [D29773962](https://our.internmc.facebook.com/intern/diff/D29773962/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D29773962/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Log if graph is static at end of training"


When set_static_graph=False, the only type of dynamism we really
support in DDP is dynamic set of unused parameters which must be explicitly
enabled with find_unused_parameters=True. Although, some workflows have static
set of unused parameters, would be good to detect and add this to logging to
identify workflows that are candidates for static graph optimization.

Differential Revision: [D29773962](https://our.internmc.facebook.com/intern/diff/D29773962/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D29773962/)!

[ghstack-poisoned]

zhaojuanmao commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -1211,6 +1216,14 @@ void Reducer::search_unused_parameters(
         "flag off. Note that this warning may be a false positive if your model "
         "has flow control causing later iterations to have unused parameters.");
   }
+  if (!static_graph_) {
```
nit: if (!static_graph_ && ddp_graph_static_), so that the {} codes can be skipped if ddp_graph_static_ is false

zhaojuanmao reviewed (approved):
one nit

rohan-varma committed:
Update on "[DDP] Log if graph is static at end of training"


When set_static_graph=False, the only type of dynamism we really
support in DDP is dynamic set of unused parameters which must be explicitly
enabled with find_unused_parameters=True. Although, some workflows have static
set of unused parameters, would be good to detect and add this to logging to
identify workflows that are candidates for static graph optimization.

Differential Revision: [D29773962](https://our.internmc.facebook.com/intern/diff/D29773962/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D29773962/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Log if graph is static at end of training"


When set_static_graph=False, the only type of dynamism we really
support in DDP is dynamic set of unused parameters which must be explicitly
enabled with find_unused_parameters=True. Although, some workflows have static
set of unused parameters, would be good to detect and add this to logging to
identify workflows that are candidates for static graph optimization.

Differential Revision: [D29773962](https://our.internmc.facebook.com/intern/diff/D29773962/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D29773962/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Log if graph is static at end of training"


When set_static_graph=False, the only type of dynamism we really
support in DDP is dynamic set of unused parameters which must be explicitly
enabled with find_unused_parameters=True. Although, some workflows have static
set of unused parameters, would be good to detect and add this to logging to
identify workflows that are candidates for static graph optimization.

Differential Revision: [D29773962](https://our.internmc.facebook.com/intern/diff/D29773962/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D29773962/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Log if graph is static at end of training"


When set_static_graph=False, the only type of dynamism we really
support in DDP is dynamic set of unused parameters which must be explicitly
enabled with find_unused_parameters=True. Although, some workflows have static
set of unused parameters, would be good to detect and add this to logging to
identify workflows that are candidates for static graph optimization.

Differential Revision: [D29773962](https://our.internmc.facebook.com/intern/diff/D29773962/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D29773962/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Log if graph is static at end of training"


When set_static_graph=False, the only type of dynamism we really
support in DDP is dynamic set of unused parameters which must be explicitly
enabled with find_unused_parameters=True. Although, some workflows have static
set of unused parameters, would be good to detect and add this to logging to
identify workflows that are candidates for static graph optimization.

Differential Revision: [D29773962](https://our.internmc.facebook.com/intern/diff/D29773962/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D29773962/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Log if graph is static at end of training"


When set_static_graph=False, the only type of dynamism we really
support in DDP is dynamic set of unused parameters which must be explicitly
enabled with find_unused_parameters=True. Although, some workflows have static
set of unused parameters, would be good to detect and add this to logging to
identify workflows that are candidates for static graph optimization.

Differential Revision: [D29773962](https://our.internmc.facebook.com/intern/diff/D29773962/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D29773962/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Log if graph is static at end of training"


When set_static_graph=False, the only type of dynamism we really
support in DDP is dynamic set of unused parameters which must be explicitly
enabled with find_unused_parameters=True. Although, some workflows have static
set of unused parameters, would be good to detect and add this to logging to
identify workflows that are candidates for static graph optimization.

Differential Revision: [D29773962](https://our.internmc.facebook.com/intern/diff/D29773962/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D29773962/)!

[ghstack-poisoned]
